### DISCUSSION TITLE
Request for Official Roadmap and Timeline for OpenTCS VDA5050 Adapter

### DISCUSSION THREAD

**Discussion by z29591259** [openTCS/opentcs#141] [General]
Dear OpenTCS Open Source Community Development Team,

My name is Peter Yu and I am a developer working on an AGV transportation system based on OpenTCS and the VDA5050 Adapter. During testing, I have encountered some issues, including:

 - Incomplete state null handling
 - Some optional fields are not yet implemented...etc
 
I plan to complete the fixes for these issues by mid-July, making the VDA5050 Adapter functionally usable to meet my needs.

I would like to inquire about the official development roadmap and timeline for the VDA5050 Adapter and explore the possibility of collaboration. I am willing to contribute code to jointly promote the development of the Adapter.

I believe that through collaboration between us, we can accelerate the development progress of the Adapter and provide users with a more stable and reliable product.

I look forward to your reply.

Sincerely,
Peter Yu

**Comment by swltr** [comment]
@z29591259 At this point, there isn't any fixed roadmap with a timeline - improvements happen as they are required in funded projects or as free time permits. But we have some improvements in the queue for the handling of optional/`null` values, which we simply haven't found the time to merge, yet. Contributions are certainly welcome! However, I think it would be best to wait approximately a couple of weeks and then coordinate to avoid potential conflicts between changes. In the meantime, maybe you could provide some details about what changes you intend to make.

**Reply by z29591259** [reply]
Currently, the modifications I have made include:

Expanded the VehicleSimulator 2.0 to validate the VDA5050 Adapter
1.1 Supports the execution of all BlockingType Actions, with the ability to trigger FATAL Errors based on a configured probability
1.2 Supports multi-vehicle operation, with the initial positions of each vehicle pre-configured
1.3 Sends State updates every 30 seconds
1.4 Does not reset positions after MQTT disconnection

Extended the Adapter
2.1 Fixed state null exception issue
2.2 Added custom logic: when the Vehicle State changes from IDLE to ERROR, it immediately cancels the task, and reads custom attributes from map to send custom actions

The next step, I will test with actual vehicles.

**Reply by swltr** [reply]
@z29591259 Thanks for your explanations, they are appreciated!

* 1.1 to 1.4: Having a way to _properly_ simulate a vehicle is something that can certainly be useful in general. However, I'd be reluctant to integrate it into this project, to be honest. The `VehicleSimulator` class here was only intended to help with testing the parsing of messages exchanged, without any logic or semantics being tested, as it would be unclear where to draw the line between this and a full vehicle-side implementation. I think this is out of scope for the driver project, and it would be better to manage it in a separate project.
* 2.1: Can you elaborate on this? A few changes have been integrated recently. Is the issue this is about still present?
* 2.2: Currently, the driver supports optionally withdrawing the transport order when the vehicle reports a change in the operating mode. When the vehicle reports an error and that error is temporary, withdrawing the transport order would seem unnecessary / overly drastic. What is your use case here, and what custom actions would you send to the vehicle?

**Comment by manjushreearadhya2016** [comment]
Hi @z29591259 ,Can you help in resolving this?
We are trying to integrate `opentcs-commadapter-vda5050` with OpenTCS version 5.17.0. We have performed the following steps:

1. Added `opentcs-commadapter-vda5050` implementation in the main `build.gradle`.
2. Added `opentcs-commadapter-vda5050` implementation in `OpenTCS-Kernel`'s `build.gradle`.
3. Added `opentcs-commadapter-vda5050` implementation in `OpenTCS-KernelControlCenter`'s `build.gradle`.
4. Added `opentcs-commadapter-vda5050` implementation in `OpenTCS-ModelEditor`'s `build.gradle`.

The build was successful, but we encountered the following error while running the kernel:

 An exception was caught and reported. Message: java.lang.IllegalArgumentException: org.opentcs.customizations.kernel.ActiveInAllModes is not a binding annotation. Please annotate it with @BindingAnnotation.
java.lang.IllegalArgumentException: org.opentcs.customizations.kernel.ActiveInAllModes is not a binding annotation. Please annotate it with @BindingAnnotation.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:220)
        at com.google.inject.Key.ensureIsBindingAnnotation(Key.java:382)
        at com.google.inject.Key.strategyFor(Key.java:370)
        at com.google.inject.Key.get(Key.java:229)
        at com.google.inject.multibindings.Multibinder.newSetBinder(Multibinder.java:146)
        at org.opentcs.customizations.kernel.KernelInjectionModule.extensionsBinderAllModes(KernelInjectionModule.java:75)
        at org.opentcs.kernel.extensions.adminwebapi.AdminWebApiModule.configure(AdminWebApiModule.java:46)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:64)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
        at com.google.inject.spi.Elements.getElements(Elements.java:113)
        at com.google.inject.util.Modules$OverrideModule.configure(Modules.java:238)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:64)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
        at com.google.inject.spi.Elements.getElements(Elements.java:113)
        at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:160)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:107)
        at com.google.inject.Guice.createInjector(Guice.java:87)
        at com.google.inject.Guice.createInjector(Guice.java:69)
        at com.google.inject.Guice.createInjector(Guice.java:59)
        at org.opentcs.kernel.RunKernel.main(RunKernel.java:60)

[20240607-11:58:51-449] INFO    main                 c.g.i.internal.MessageProcessor.visit()                : An exception was caught and reported. Message: java.lang.IllegalArgumentException: org.opentcs.customizations.kernel.ActiveInAllModes is not a binding annotation. Please annotate it with @BindingAnnotation.
java.lang.IllegalArgumentException: org.opentcs.customizations.kernel.ActiveInAllModes is not a binding annotation. Please annotate it with @BindingAnnotation.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:220)
        at com.google.inject.Key.ensureIsBindingAnnotation(Key.java:382)
        at com.google.inject.Key.strategyFor(Key.java:370)
        at com.google.inject.Key.get(Key.java:229)
        at com.google.inject.multibindings.Multibinder.newSetBinder(Multibinder.java:146)
        at org.opentcs.customizations.kernel.KernelInjectionModule.extensionsBinderAllModes(KernelInjectionModule.java:75)
        at org.opentcs.kernel.extensions.servicewebapi.ServiceWebApiModule.configure(ServiceWebApiModule.java:46)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:64)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
        at com.google.inject.spi.Elements.getElements(Elements.java:113)
        at com.google.inject.util.Modules$OverrideModule.configure(Modules.java:238)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:64)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
        at com.google.inject.spi.Elements.getElements(Elements.java:113)
        at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:160)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:107)
        at com.google.inject.Guice.createInjector(Guice.java:87)
        at com.google.inject.Guice.createInjector(Guice.java:69)
        at com.google.inject.Guice.createInjector(Guice.java:59)
        at org.opentcs.kernel.RunKernel.main(RunKernel.java:60)

[20240607-11:58:51-450] INFO    main                 c.g.i.internal.MessageProcessor.visit()                : An exception was caught and reported. Message: java.lang.IllegalArgumentException: org.opentcs.customizations.kernel.ActiveInAllModes is not a binding annotation. Please annotate it with @BindingAnnotation.
java.lang.IllegalArgumentException: org.opentcs.customizations.kernel.ActiveInAllModes is not a binding annotation. Please annotate it with @BindingAnnotation.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:220)
        at com.google.inject.Key.ensureIsBindingAnnotation(Key.java:382)
        at com.google.inject.Key.strategyFor(Key.java:370)
        at com.google.inject.Key.get(Key.java:229)
        at com.google.inject.multibindings.Multibinder.newSetBinder(Multibinder.java:146)
        at org.opentcs.customizations.kernel.KernelInjectionModule.extensionsBinderAllModes(KernelInjectionModule.java:75)
        at org.opentcs.kernel.extensions.rmi.RmiServicesModule.configure(RmiServicesModule.java:83)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:64)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
        at com.google.inject.spi.Elements.getElements(Elements.java:113)
        at com.google.inject.util.Modules$OverrideModule.configure(Modules.java:238)
        at com.google.inject.AbstractModule.configure(AbstractModule.java:64)
        at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
        at com.google.inject.spi.Elements.getElements(Elements.java:113)
        at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:160)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:107)
        at com.google.inject.Guice.createInjector(Guice.java:87)
        at com.google.inject.Guice.createInjector(Guice.java:69)
        at com.google.inject.Guice.createInjector(Guice.java:59)
        at org.opentcs.kernel.RunKernel.main(RunKernel.java:60)



Please guide us on resolving these issues and completing the integration of `opentcs-commadapter-vda5050`.

Thank you.

**Comment by ptiwari0664** [comment]
I am facing the same issues and not able to integrate the adapter with 5.19.0 kernel. @z29591259  @swltr  Can you please share detailed integration document ?

**Comment by martingr** [comment]
@manjushreearadhya2016 @ptiwari0664 for further discussions on the issue you both have, please use #147.

**Comment by Gosling1702** [comment]
Hi Stefan Walter(https://github.com/swltr), Martin Grzenia(https://github.com/martingr)

I am working on the openTCS-integration-Example and I tried to integrated it with VDA5050 adapter v2.0. After Kernel, Model Editor, Operational Desk, Kernel Control Center service turned UP. I am getting in openTCS template VDA Adapter for Operational Desk as-
![image](https://github.com/openTCS/opentcs/assets/172021241/3aa46970-543a-4c24-a749-3299da762ebd)


While I enable the VDA5050 adapter, the Vehicle state is still "Unknown". and My Control Center is-
![image](https://github.com/openTCS/opentcs/assets/172021241/77c43170-285d-44fe-a93f-7dc0662d9b2c)


**my topic is :** uagv/v2/xyz/0001/state
**This is the details I send through MQTTx-**
{"headerId":11181,"timestamp":"2024-06-18T12:14:44.6046930Z","version":"2.0.0","manufacturer":"xyz","serialNumber":"0001","orderId":"","orderUpdateId":0,"lastNodeId":"","lastNodeSequenceId":0,"nodeStates":[],"edgeStates":[],"agvPosition":{"positionInitialized":true,"localizationScore":1.0,"x":-29.0,"y":-16.999998092651368,"theta":0.00019965189858339727,"mapId":"01"},"velocity":{"vx":0.0,"vy":0.0,"omega":0.0},"loads":[],"driving":false,"paused":false,"newBaseRequest":false,"distanceSinceLastNode":0.0,"actionStates":[],"batteryState":{"batteryCharge":100.0,"batteryVoltage":48.0,"batteryHealth":100.0,"charging":false},"operatingMode":"AUTOMATIC","errors":[],"information":[],"safetyState":{"eStop":"NONE","fieldViolation":false}}

### It will connected with broker, but not with the Vehicle. Guide me like how can I enable Vehicle connection.

**Reply by z29591259** [reply]
You need to check the message in topic `uagv/v2/xyz/0001/connection` is online. Then `Vehicle connected` shoud turn to green

**Reply by Gosling1702** [reply]
Thank you for Replying @z29591259, I am facing **issue in setting the location of Vehicle** in the operational Desk. After enable the vehicle(from Kernel Control Center), how can i set the position of vehicle in the Operational Desk?
![image](https://github.com/openTCS/opentcs/assets/172021241/612f56d3-327d-4495-87d1-21a6d1693509)
Here, I am not be able to set the Vehicle Position. It is showing null for me.

**Comment by Gosling1702** [comment]
Hello @z29591259, @martingr, @swltr,

In opentcs-Integration-Example, I integrated the Vehicle with the VDA5050 Adapter and now, it is connected with Broker and Vehicle as well.

After enable the vehicle from Kernel Control Center, I am **not be able to set the initial position of the Vehicle in the Operation Desk.**
![image](https://github.com/openTCS/opentcs/assets/172021241/5d2207fd-e6a9-4c6b-9e7d-97f8b268074d)

![image](https://github.com/openTCS/opentcs/assets/172021241/f6fd722a-c8e0-4676-a865-1bdc152bddd1)

Now, **How can i set the position of the Vehicle, **position of the Vehicle is showing me Null****. 

### How can we initially move the vehicle from some random location to the nearest Point in the opentcs-Integration-Example using VDA5050 adapter.

After giving the order to the vehicle, it is also going the random place for me and sometime it is directly reach to the destination.

**Reply by z29591259** [reply]
1. Because vehicle position in OperationDesk is reference point xy model value, you could execute "Copy model values to layou" in menu "Actions", vehicle icon position on UI should be correct.

![image](https://github.com/openTCS/opentcs/assets/25024958/6ceff673-5214-42d5-93a1-77749d2a49b2)

2. Check the message of Agv state topic, lastNodeId should not be empty or null.

**Reply by Gosling1702** [reply]
Thank you @z29591259,

Now, I am able to set the vehicle position, But **in my UI, the path of the vehicle is not update(red dotted lines)** When the Vehicle starts moving from One point to another point, the Red Dotted lines are not updating. And when the Vehicle reach at the Destination, the Order is still in progress and it is not processing the another order, because the previous order is not complete(Order is still in progress).

**Reply by z29591259** [reply]
The resource update (red dotted lines) and order completion depend on the `nodeStates` and `edgeStates` in `uagv/v2/xyz/0001/state`. 
Ensure that the AGV state adheres to the [VDA5050 flow](https://github.com/VDA5050/VDA5050/blob/main/VDA5050_EN.md#-6102-traversal-of-nodes-and-enteringleaving-edges-triggering-of-actions).

**Comment by Sadhu17** [comment]
Hi Peter, [z29591259](https://github.com/z29591259)

I am integrating the VDA5050 version 2.0 adapter with Opentcs-5.11.0 and encountering some issues. The Vehicle and Broker connect successfully at the KernelControlCenter, but problems arise when I try to move the vehicle at the OperationalDesk. I would appreciate any guidance on how to resolve the following issues:

1.- After enabling the vehicle and changing its integration level to "Integrated Fully", the vehicle moves erratically and in reverse.
2.- The vehicle’s movement path is not updating correctly, and previously traversed paths are not being cleared.
3.- Once the vehicle reaches its destination, the order status remains "Being Processed," preventing the vehicle from processing the next queued order.
4.- Is there an alternative method for assigning multiple orders to the vehicle?
5.- The message containing "orderId": "TOrder-01J1YQG11PPD8FN7HWNCTZC6ZH-0" and "orderUpdateId": 1 is being published continuously same one like {orderId": "TOrder-01J1YQG11PPD8FN7HWNCTZC6ZH-0" and "orderUpdateId"}, but ideally it has to be published once?

I am also including my simulation topic and message for reference.

Topic: uagv/v2/XYZ/0001/order

Message{
  "headerId" : 162,
  "timestamp" : "2024-07-04T11:24:24.247443300Z",
  "version" : "2.0.0",
  "manufacturer" : "XYZ",
  "serialNumber" : "0001",
  "orderId" : "TOrder-01J1YQG11PPD8FN7HWNCTZC6ZH-0",
  "orderUpdateId" : 1,
  "nodes" : [ {
    "nodeId" : "P-0061",
    "sequenceId" : 2,
    "released" : true,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -15.74,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0009",
    "sequenceId" : 4,
    "released" : true,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -13.448,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0063",
    "sequenceId" : 6,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0067",
    "sequenceId" : 8,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -15.5,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0071",
    "sequenceId" : 10,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -13.64,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0075",
    "sequenceId" : 12,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -11.82,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0079",
    "sequenceId" : 14,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -10.05,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0087",
    "sequenceId" : 16,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -8.36,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0006",
    "sequenceId" : 18,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -8.36,
      "y" : -7.0,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0023",
    "sequenceId" : 20,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -8.36,
      "y" : -1.52,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0022",
    "sequenceId" : 22,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -12.68,
      "y" : -1.52,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0038",
    "sequenceId" : 24,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -1.5,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0021",
    "sequenceId" : 26,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -21.33,
      "y" : -1.52,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0007",
    "sequenceId" : 28,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -25.6,
      "y" : -1.5,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0008",
    "sequenceId" : 30,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -25.6,
      "y" : 2.5,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0025",
    "sequenceId" : 32,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -21.33,
      "y" : 2.51,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0029",
    "sequenceId" : 34,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.54,
      "y" : 2.53,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0037",
    "sequenceId" : 36,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.54,
      "y" : 5.3,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0039",
    "sequenceId" : 38,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -14.47,
      "y" : 5.3,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0040",
    "sequenceId" : 40,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -14.47,
      "y" : 8.09,
      "mapId" : ""
    }
  } ],
  "edges" : [ {
    "edgeId" : "P-0061 --- P-0009",
    "sequenceId" : 3,
    "released" : true,
    "startNodeId" : "P-0061",
    "endNodeId" : "P-0009",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0009 --- P-0063",
    "sequenceId" : 5,
    "released" : false,
    "startNodeId" : "P-0009",
    "endNodeId" : "P-0063",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0063 --- P-0067",
    "sequenceId" : 7,
    "released" : false,
    "startNodeId" : "P-0063",
    "endNodeId" : "P-0067",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0067 --- P-0071",
    "sequenceId" : 9,
    "released" : false,
    "startNodeId" : "P-0067",
    "endNodeId" : "P-0071",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0071 --- P-0075",
    "sequenceId" : 11,
    "released" : false,
    "startNodeId" : "P-0071",
    "endNodeId" : "P-0075",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0075 --- P-0079",
    "sequenceId" : 13,
    "released" : false,
    "startNodeId" : "P-0075",
    "endNodeId" : "P-0079",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0079 --- P-0087",
    "sequenceId" : 15,
    "released" : false,
    "startNodeId" : "P-0079",
    "endNodeId" : "P-0087",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0087 --- P-0006",
    "sequenceId" : 17,
    "released" : false,
    "startNodeId" : "P-0087",
    "endNodeId" : "P-0006",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0006 --- P-0023",
    "sequenceId" : 19,
    "released" : false,
    "startNodeId" : "P-0006",
    "endNodeId" : "P-0023",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0023 --- P-0022",
    "sequenceId" : 21,
    "released" : false,
    "startNodeId" : "P-0023",
    "endNodeId" : "P-0022",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0022 --- P-0038",
    "sequenceId" : 23,
    "released" : false,
    "startNodeId" : "P-0022",
    "endNodeId" : "P-0038",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0038 --- P-0021",
    "sequenceId" : 25,
    "released" : false,
    "startNodeId" : "P-0038",
    "endNodeId" : "P-0021",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0021 --- P-0007",
    "sequenceId" : 27,
    "released" : false,
    "startNodeId" : "P-0021",
    "endNodeId" : "P-0007",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0007 --- P-0008",
    "sequenceId" : 29,
    "released" : false,
    "startNodeId" : "P-0007",
    "endNodeId" : "P-0008",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0008 --- P-0025",
    "sequenceId" : 31,
    "released" : false,
    "startNodeId" : "P-0008",
    "endNodeId" : "P-0025",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0025 --- P-0029",
    "sequenceId" : 33,
    "released" : false,
    "startNodeId" : "P-0025",
    "endNodeId" : "P-0029",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0029 --- P-0037",
    "sequenceId" : 35,
    "released" : false,
    "startNodeId" : "P-0029",
    "endNodeId" : "P-0037",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0037 --- P-0039",
    "sequenceId" : 37,
    "released" : false,
    "startNodeId" : "P-0037",
    "endNodeId" : "P-0039",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0039 --- P-0040",
    "sequenceId" : 39,
    "released" : false,
    "startNodeId" : "P-0039",
    "endNodeId" : "P-0040",
    "actions" : [ ],
    "maxSpeed" : 1.0
  },

**Reply by z29591259** [reply]
This Order JSON is being sent for the first time and only includes two base nodes. The key is how the AGV's state responds. 

The AGV's state should update the orderId and orderUpdateId according to the [VDA5050 flow](https://github.com/VDA5050/VDA5050/blob/main/VDA5050_EN.md#-6102-traversal-of-nodes-and-enteringleaving-edges-triggering-of-actions) before it can start moving. Then, the TCS will continue to send new orderUpdates to the AGV.

**Reply by Sadhu17** [reply]
Hello [z29591259](https://github.com/z29591259), [martingr](https://github.com/martingr), [swltr](https://github.com/swltr)

While I am running the openTCS 5.11.0 using VDA5050 Adapter v2.0, My "order" Topic and message publish multiple time, but ideally it should publish once? 
 Topic and Message are as, **"orderId" : **"TOrder-01J2B6VWFRJXEZ4J8CXD1PYWRP-0**", "orderUpdateId" : **1****, was published multiple time with the Same data, How Can I resolve this?

My Vehicle is moving in the reverse direction, and previous traversed path is not updated. I am attaching the images to understand-

![Picture2](https://github.com/openTCS/opentcs/assets/174597698/fd66a526-2694-42b0-b1be-c3bc846d7836)
![Picture1](https://github.com/openTCS/opentcs/assets/174597698/72c8f06d-b286-42bc-975f-dc986d712c10)

My order Topic is: uagv/v2/XYZ/0001/order
Message {
  "headerId" : 143,
  "timestamp" : "2024-07-09T07:40:48.620412800Z",
  "version" : "2.0.0",
  "manufacturer" : "XYZ",
  "serialNumber" : "0001",
  "orderId" : **"TOrder-01J2B6VWFRJXEZ4J8CXD1PYWRP-0**",
  "orderUpdateId" : **1**,
  "nodes" : [ {
    "nodeId" : "P-0059",
    "sequenceId" : 2,
    "released" : true,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -5.5,
      "y" : 5.3,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0028",
    "sequenceId" : 4,
    "released" : true,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -5.5,
      "y" : 2.51,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0024",
    "sequenceId" : 6,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -5.5,
      "y" : -1.52,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0023",
    "sequenceId" : 8,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -8.36,
      "y" : -1.52,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0022",
    "sequenceId" : 10,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -12.68,
      "y" : -1.52,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0038",
    "sequenceId" : 12,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -1.5,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0005",
    "sequenceId" : 14,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -7.0,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0063",
    "sequenceId" : 16,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -10.81,
      "mapId" : ""
    }
  }, {
    "nodeId" : "P-0009",
    "sequenceId" : 18,
    "released" : false,
    "actions" : [ ],
    "nodePosition" : {
      "x" : -17.25,
      "y" : -13.448,
      "mapId" : ""
    }
  } ],
  "edges" : [ {
    "edgeId" : "P-0059 --- P-0028",
    "sequenceId" : 3,
    "released" : true,
    "startNodeId" : "P-0059",
    "endNodeId" : "P-0028",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0028 --- P-0024",
    "sequenceId" : 5,
    "released" : false,
    "startNodeId" : "P-0028",
    "endNodeId" : "P-0024",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0024 --- P-0023",
    "sequenceId" : 7,
    "released" : false,
    "startNodeId" : "P-0024",
    "endNodeId" : "P-0023",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0023 --- P-0022",
    "sequenceId" : 9,
    "released" : false,
    "startNodeId" : "P-0023",
    "endNodeId" : "P-0022",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0022 --- P-0038",
    "sequenceId" : 11,
    "released" : false,
    "startNodeId" : "P-0022",
    "endNodeId" : "P-0038",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0038 --- P-0005",
    "sequenceId" : 13,
    "released" : false,
    "startNodeId" : "P-0038",
    "endNodeId" : "P-0005",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0005 --- P-0063",
    "sequenceId" : 15,
    "released" : false,
    "startNodeId" : "P-0005",
    "endNodeId" : "P-0063",
    "actions" : [ ],
    "maxSpeed" : 1.0
  }, {
    "edgeId" : "P-0063 --- P-0009",
    "sequenceId" : 17,
    "released" : false,
    "startNodeId" : "P-0063",
    "endNodeId" : "P-0009",
    "actions" : [ ],
    "maxSpeed" : 1.0
  } ]
}

and state message as : State{header=Header{headerId=9792, timestamp=2024-07-09T07:40:47.034525Z, version=2.0.0, manufacturer=xyz, serialNumber=0001}, orderId=TOrder-01J2B6VWFRJXEZ4J8CXD1PYWRP-0, orderUpdateId=0, zoneSetId=null, lastNodeId=P-0009, lastNodeSequenceId=18, nodeStates=[], edgeStates=[], agvPosition=AgvPosition{x=-17.25, y=-13.4399995803833, theta=1.5728907585144043, mapId=0, mapDescription=null, positionInitialized=true, localizationScore=1.0, deviationRange=0.0}, velocity=Velocity{vx=0.0, vy=0.0, omega=0.0}, loads=[], driving=false, paused=false, newBaseRequest=false, distanceSinceLastNode=0.0, actionStates=[], batteryState=BatteryState{batteryCharge=100.0, batteryVoltage=48.0, batteryHealth=100, charging=false, reach=null}, operatingMode=AUTOMATIC, errors=[], information=[], safetyState=SafetyState{eStop=NONE, fieldViolation=false}}

Now, I am not able to figure it out.

**Reply by swltr** [reply]
@Sadhu17 Please do not hijack discussion topics with your own questions. but stick to creating your own topic.
